c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -580,12 +593,51 @@ impl<'a, F: FnMut(&str)> Sink for RewriteTreeSink<'a, F> {
     }
 }
 
+#[derive(Debug, Default)]
+struct LineMapBuilder {
+    /// Map from line indices in the original source code to line indices in `buf`.  For each input
+    /// line, we give the index of the first output line containing some unmodified portion of the
+    /// input line.
+    v: Vec<Option<usize>>,
+}
+
+impl LineMapBuilder {
+    /// Record that part of input line `i` is found in output line `j`.
+    pub fn record(&mut self, i: usize, j: usize) {
+        if i >= self.v.len() {
+            self.v.resize(i + 1, None);
+        }
+        self.v[i] = match self.v[i] {
```

Comment:
`self.v[i] = Some(cmp::min(j, self.v[i].unwrap_or(j)));` or something like that would be shorter.